### PR TITLE
Support rdc device funcs

### LINKED ISSUE #40: Jitting kernels that call device functions defined in different translation unit breaks proteus
When JIT GPU kernels which call a device function implemented in a different translation unit proteus fails.  PR: #39 tests for this case and currently is marked as `WILL_FAIL`.

### LINKED ISSUE #45: Refactor pass implementation
The current pass implementation is hard to parse. We should implement some class at least that reflects the expectation of the runtime library.

### LINKED ISSUE #46: Tests on tioga for rocm 6.2.1/LLVM 18 fail
Test failure on rocm 6.2.1 on tioga:
```
The following tests FAILED:
	 21 - kernel_host_device_jit.HIP (Failed)
	 38 - kernel_host_device_jit.HIP.rdc (Failed)
```

The `kernel_host_device_jit` test creates a host JIT function that calls a kernel. The issues are:
1. In LLVM 18, ORC JIT does not make globals on custom sections externally visible
2. In LLVM 18, HIP creates a mangled global (`__hip_fatbin_<hash>`) to store the hip fatbin when Proteus using ORC JIT expects the global to be named simply `__hip_fatbin`.

We expect to solve that moving fatbin extraction through instrumentation. Relates to #39, #40

### PR DESCRIPTION
**Pull request by ggeorgakoudis**
Closes #40 
Closes #45
Closes #46

### CODE REVIEW

File: lib/JitEngineDevice.hpp [hunk 1]
```diff
@@ -78,21 +82,31 @@ template <typename ImplT> class JitEngineDevice : protected JitEngine {
...
+    SmallVector<std::string> LinkedModuleIds;
+  };
 
-  void insertRegisterFunction(const void *HostAddr, char *FunctionName,
-                              int32_t *RCIndices, int32_t NumRCs) {
-    JITKernelInfoMap[HostAddr] = JITKernelInfo(FunctionName, RCIndices, NumRCs);
-  }
+  void *CurHandle = nullptr;
+  DenseMap<void *, BinaryInfo> HandleToBinaryInfo;
+  DenseMap<void *, void *> KernelToHandleMap;
```
Comment (koparasy) [inline lib/JitEngineDevice.hpp:108]: Maybe better name.

File: lib/JitEngineDeviceHIP.cpp [hunk 2]
```diff
@@ -162,6 +208,16 @@ JitEngineDeviceHIP::codegenObject(Module &M, StringRef DeviceArch) {
...
+#if 0
+  for (auto &KV : LinkedModules) {
+    auto DeviceBitcode = KV.second;
+    hiprtcErrCheck(
+        hiprtcLinkAddData(hip_link_state_ptr, HIPRTC_JIT_INPUT_LLVM_BITCODE,
+                          (void *)DeviceBitcode.data(), DeviceBitcode.size(),
+                          "", 0, nullptr, nullptr));
+  }
+#endif
+
```
Comment (koparasy) [inline lib/JitEngineDeviceHIP.cpp:220]: Delete.

File: pass/ProteusPass.cpp [hunk 3]
```diff
@@ -133,27 +132,32 @@ class ProteusJitPassImpl {
     // Host compilation
     // ================
 
+    instrumentRegisterLinkedBinary(M);
+    instrumentRegisterFatBinary(M);
+    instrumentRegisterFatBinaryEnd(M);
+    instrumentRegisterVar(M);
```
Comment (koparasy) [inline pass/ProteusPass.cpp:138]: We don't have a test for this yet.

File: pass/ProteusPass.cpp [hunk 4]
```diff
@@ -133,27 +132,32 @@ class ProteusJitPassImpl {
...
         continue;
-      }
 
       emitJitModuleHost(M, JFI);
       emitJitEntryCall(M, JFI);
     }
 
+    DEBUG(dbgs() << "=== Post Original Host Module\n"
+                 << M << "=== End Post Original Host Module\n");
+
```
Comment (koparasy) [inline pass/ProteusPass.cpp:160]: Remove.

File: pass/ProteusPass.cpp [hunk 5]
```diff
@@ -738,37 +637,20 @@ class ProteusJitPassImpl {
...
-    GlobalVariable *FatbinWrapper = nullptr;
-    FatbinWrapper = M.getGlobalVariable(FatbinName, true);
-    if (!FatbinWrapper)
-      FATAL_ERROR(
-          "Expected hip fatbinary wrapper global, check "
-          "ENABLE_CUDA|ENABLE_HIP compilation flags for ProteusJitPass");
-
-    FunctionType *JitEntryFnTy = nullptr;
+  FunctionCallee getJitLaunchKernelFn(Module &M) {
+    FunctionType *JitLaunchKernelFnTy = nullptr;
```
Comment (koparasy) [inline pass/ProteusPass.cpp:641]: Since this is a refactor once more. I would add a TODO/FIXME note here and mention the ABI issue

File: pass/ProteusPass.cpp (hunk 3 above)
Comment (koparasy) [inline pass/ProteusPass.cpp:138]: This needs to be guarded for code that contains device stubs.

Comment (koparasy) [reply pass/ProteusPass.cpp:138]: I checked also the function `hasDeviceLaunchKernelCalls` and is not sufficient. You need to instrument when __hip/cuda_module_ctor is being called.

File: pass/ProteusPass.cpp [hunk 6]
```diff
@@ -868,6 +742,135 @@ class ProteusJitPassImpl {
...
+        "__jit_register_fatbinary", JitRegisterFatbinaryFnTy);
+
+    return JitRegisterFatbinaryFn;
+  }
+
+  void instrumentRegisterFatBinary(Module &M) {
+    Function *F = nullptr;
+
+    if (!RegisterFatBinaryName)
+      return;
```
Comment (koparasy) [inline pass/ProteusPass.cpp:759]: This is a convoluted logic. Isn't there a better way to skip this code? Or, better never call this function if there is nothing todo.

File: pass/ProteusPass.cpp [hunk 7]
```diff
@@ -868,6 +742,135 @@ class ProteusJitPassImpl {
...
+
+    F = M.getFunction(RegisterFatBinaryName);
+    if (!F)
+      return;
+
+    FunctionCallee JitRegisterFatBinaryFn = getJitRegisterFatBinaryFn(M);
```
Comment (koparasy) [inline pass/ProteusPass.cpp:765]: In the IRs that I am looking at this call is never being made when enabling rdc. It is only generated by the link.stub. I think you need access to the individual models (fatbins) as well.

File: pass/ProteusPass.cpp [hunk 8]
```diff
@@ -868,6 +742,135 @@ class ProteusJitPassImpl {
...
+    // link.stub. It's not strictly necessary since this module will not have a
+    // device bitcode to pull and we skip at runtime.
+    if (!M.getGlobalVariable("__cuda_fatbin_wrapper", /*AllowInternal=*/true)) {
+      DEBUG(dbgs() << "Skip " << M.getSourceFileName() << "\n";)
+      return;
+    }
+
+    FunctionCallee JitRegisterLinkedBinaryFn = getJitRegisterLinkedBinaryFn(M);
+
+    for (auto &F : M.getFunctionList()) {
```
Comment (koparasy) [inline pass/ProteusPass.cpp:843]: Instead of iterating all functions, get the handle `__cuda_module_ctor`. Inside that function __cudaRegisterLinkedBinary is being called. The constructor is small, so in principle it should be faster.

File: pass/ProteusPass.cpp [hunk 9]
```diff
@@ -868,6 +742,135 @@ class ProteusJitPassImpl {
...
+        if (!CB)
+          continue;
+
+        IRBuilder<> Builder(CB);
+        std::string GVName = getJitBitcodeUniqueName(M);
+        DEBUG(
+            dbgs() << "Instrument register linked binary to extract bitcode GV "
+                   << GVName << "\n");
+        auto *Arg = Builder.CreateGlobalString(GVName);
+        Builder.CreateCall(JitRegisterLinkedBinaryFn, {Arg});
```
Comment (koparasy) [inline pass/ProteusPass.cpp:858]: This will be deleted by entering the ctor body.

File: tests/gpu/device_func.cpp [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#include "gpu_common.h"
+#include <cstdio>
+
+__device__ void device_function(int a) { printf("device_func %d\n", a); }
+
+__global__ void kernel2() { printf("Kernel2\n"); }
```
Comment (koparasy) [inline tests/gpu/device_func.cpp:6]: Either delete `kernel2` or create another test that has only device functions. 

The inclusion of a kernel, will enable your pass to perform additional actions (e.g. there is kernel stub and it will instrument it). But we need to test the case that there is no such stub.

Comment (koparasy) [reply tests/gpu/device_func.cpp:6]: Can you also change:

`__device__ void device_function(int a) { printf("device_func %d\n", a); } ` 

to 

__device__ void device_function(int *a) {*a  = 123;}

Thus avoiding the TU to contain an external function?

**Review by koparasy — CHANGES_REQUESTED** [review]
This overall looks good. The runtime portion is pretty clean. The pass requires work clean up.

1. Fix the comment on the tests.
2. I suggest we take advantage of the fact that clang will always insert a call `__vendor_module_ctor` that will be called by the host if the device side contains anything interesting for us. I would use this to clean up the additions. (e.g. you can detect whether you need to perform host-register related actions based on the existence of the function).

You don't need to fix in this PR the issue with external functions. But fixing 2 should (in principle) pave the way to that fix.

Thank you

File: lib/JitEngineDevice.hpp (hunk 1 above)
Comment (ggeorgakoudis) [reply lib/JitEngineDevice.hpp:108]: Defer for later refactoring

File: pass/ProteusPass.cpp (hunk 5 above)
Comment (ggeorgakoudis) [reply pass/ProteusPass.cpp:641]: There is an issue #47

File: pass/ProteusPass.cpp (hunk 6 above)
Comment (ggeorgakoudis) [reply pass/ProteusPass.cpp:759]: that's our alternative for for #ifdefs. Defer, refactoring

File: tests/gpu/device_func.cpp (hunk 10 above)
Comment (ggeorgakoudis) [reply tests/gpu/device_func.cpp:6]: Why avoid the external function? I'm using it to check using lit

File: pass/ProteusPass.cpp (hunk 7 above)
Comment (ggeorgakoudis) [reply pass/ProteusPass.cpp:765]: Proteus compiles the link.stub. I'll consider it resolved.

File: pass/ProteusPass.cpp (hunk 8 above)
Comment (ggeorgakoudis) [reply pass/ProteusPass.cpp:843]: This is a good suggestion. I'll defer for iteration 2 and open an issue.

**Review by koparasy — APPROVED** [review]
